File: 이준영/프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;
class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
		Map<String, Boolean> map = new HashMap<>();
		int len = begin.length();
		for (String str : words) {
			map.put(str, false);
		}
		map.put(begin, false);
		if (!map.containsKey(target)) {
			return 0;
		}
		Queue<String> queue = new LinkedList<>();
		queue.add(begin);
		while (!queue.isEmpty()) {
			answer += 1;
			int count = queue.size();
			for (int i = 0; i < count; i++) {
				String str = queue.poll();
				if (!map.get(str)) {
                    map.put(str, true);
					for (int j = 0; j < len; j++) {
						for (int k = 0; k < 26; k++) {
							StringBuilder sb = new StringBuilder(str);
							sb.setCharAt(j, (char) ('a' + k));
							String tmp = sb.toString();
							if (map.containsKey(tmp) && !map.get(tmp)) {
								queue.offer(tmp);
								if (tmp.equals(target)) {
									return answer;
								}
							}
						}
					}
				}
			}
		}
		return 0;
    }
}

File: JavaScript/이승연/프로그래머스/3/43163. 단어 변환/단어 변환.js
let min = Number.MAX_SAFE_INTEGER;

function solution(begin, target, words) {
    const map = {};
    
    words.push(begin);
    
    for (let i = 0; i < words.length; i++) {
        map[words[i]] = [];
        
        for (let j = 0; j < words.length; j++) {
            let count = 0;
            
            for (let k = 0; k < words[j].length; k++) {
                
                if (words[i][k] == words[j][k]) {
                    count++;
                }
            }
            
            if (count == (words[i].length - 1) && words[j] !== begin) {
                map[words[i]].push([words[j], 0]);
            }
        }
    }
    
    bfs(begin, target, map, 0);
    
    return min === Number.MAX_SAFE_INTEGER ? 0 : min;
}

function bfs(word, target, map, count) {
    if (word === target) {
        min = Math.min(min, count);
        // console.log(min);
        return;
    }
    
    if (count >= min) {
        return;
    }
    
    for (let i = 0; i < map[word].length; i++) {
        if (map[word][i][1] === 0) {
            map[word][i][1] = 1;
            bfs(map[word][i][0], target, map, count + 1);
            map[word][i][1] = 0;
        }
    }
}